### PR TITLE
cxx-qt-gen: remove the Data struct and move properties to Rust

### PR DESCRIPTION
**Pull request by ahayzen-kdab**

### CODE REVIEW

File: cxx-qt-gen/src/gen_cpp.rs [hunk 1]
```diff
@@ -346,6 +346,8 @@ fn generate_properties_cpp(
         let ident_getter = property.getter.as_ref().unwrap().cpp_ident.to_string();
         let ident_setter = property.setter.as_ref().unwrap().cpp_ident.to_string();
         let ident_changed = property.notify.as_ref().unwrap().cpp_ident.to_string();
+        // TODO: share in naming diff
+        let ident_emit = format!("emit{}", ident_changed.to_case(Case::Pascal));
```
Comment (ahayzen-kdab) [inline cxx-qt-gen/src/gen_cpp.rs:350]: if the naming branch lands first, we need to change this to use it.

File: examples/qml_minimal/src/cxxqt_object.rs [hunk 2]
```diff
@@ -15,36 +15,37 @@ mod ffi {
     unsafe extern "C++" {
         include!("cxx-qt-lib/include/qt_types.h");
         type QString = cxx_qt_lib::QString;
+        type QPoint = cxx_qt_lib::QPoint;
     }
 
     // ANCHOR: book_data_struct
-    pub struct Data {
+    // TODO: remove book links
```
Comment (ahayzen-kdab) [inline examples/qml_minimal/src/cxxqt_object.rs:22]: still need to update the book

File: examples/qml_minimal/src/cxxqt_object.rs [hunk 3]
```diff
@@ -15,36 +15,37 @@ mod ffi {
...
+    // ANCHOR_END: book_data_struct
+
+    // ANCHOR: book_rustobj_struct
+    #[cxx_qt::qobject]
+    pub struct MyObject {
+        #[qproperty]
         number: i32,
+        #[qproperty]
         string: UniquePtr<QString>,
```
Comment (ahayzen-kdab) [inline examples/qml_minimal/src/cxxqt_object.rs:31]: Having an opaque type C++ type stored inside a UniquePtr in the struct then causes the build to fail as `Opaque types are currently required to be [Sized](https://doc.rust-lang.org/std/marker/trait.Sized.html) and [Unpin](https://doc.rust-lang.org/std/marker/trait.Unpin.html).` https://cxx.rs/extern-rust.html#opaque-rust-types and the QString is opaque so uses PhantonPinned which doesn't implement Unpin ;-/

Comment (ahayzen-kdab) [reply examples/qml_minimal/src/cxxqt_object.rs:31]: Using `string: Box<UniquePtr<QString>>` looks like it might work :thinking:

Comment (ahayzen-kdab) [reply examples/qml_minimal/src/cxxqt_object.rs:30]: It does, but how do we express this to the developer. Or do they need to write `Box<UniquePtr<T>>` when they want to have T as a property type ? And how do we know that this type needs special conversions to T for C++ ? And what happens to the getters/setters on the Rust side? Do they return `Box<UniquePtr<T>>` ? etc

Then do we have a special type such as `Owned<T>` or something?

Comment (ahayzen-kdab) [reply examples/qml_minimal/src/cxxqt_object.rs:30]: Solved with https://github.com/dtolnay/cxx/pull/1091

File: cxx-qt/src/lib.rs [hunk 4]
```diff
@@ -118,13 +116,11 @@ pub fn qobject(_args: TokenStream, _input: TokenStream) -> TokenStream {
 /// #[cxx_qt::bridge]
 /// mod my_object {
 ///     #[derive(Default)]
-///     struct Data {
+///     struct RustObj {
```
Comment (LeonMatthesKDAB) [inline cxx-qt/src/lib.rs:119]: This is missing a `#[cxx_qt::qobject]` attribute

File: cxx-qt/src/lib.rs [hunk 5]
```diff
@@ -16,13 +16,11 @@ use cxx_qt_gen::{extract_qobject, generate_qobject_rs};
 /// #[cxx_qt::bridge(namespace = "cxx_qt::my_object")]
 /// mod my_object {
 ///     #[derive(Default)]
-///     struct Data {
+///     struct RustObj {
```
Comment (LeonMatthesKDAB) [inline cxx-qt/src/lib.rs:19]: This is missing a #[cxx_qt::qobject] attribute

**Review by LeonMatthesKDAB — COMMENTED** [review]
Noticed a few bits missing

File: cxx-qt/src/lib.rs (hunk 4 above)
Comment (ahayzen-kdab) [reply cxx-qt/src/lib.rs:121]: Done

File: cxx-qt/src/lib.rs (hunk 5 above)
Comment (ahayzen-kdab) [reply cxx-qt/src/lib.rs:20]: Done

**Review by LeonMatthesKDAB — APPROVED** [review]
LGTM

**Pull request merged**